### PR TITLE
Don't lower-case the keys in config.yaml

### LINKED ISSUE #1036: config.yaml cannot contain agent.config.createOptions since config crate lowercases field names
`struct DockerConfig` is set up so that its serde Deserializer expects a field named `createOptions`. However the config crate's YAML parser normalizes all keys to lower case. So the serde Deserializer only sees a key named `createoptions`, which it discards as unknown. The actual `create_options` field is then set to an empty default.

Even if it could be parsed, fields inside the `ContainerCreateBody` also have mixed-case names like `Hostname`, which would not match lowercased names like `hostname` provided by the config crate.

There is [an open issue](https://github.com/mehcode/config-rs/issues/65) on the config crate about this with no resolution.

We *do* want to continue using the crate since we rely on it to merge defaults + config file + env vars. While the `config::File` source's behavior of lowercasing keys is not tweakable, it should be possible to add our own YAML file [source](https://docs.rs/config/0.8.0/config/trait.Source.html) that does not lowercase the keys and use that instead.

(The lower-casing behavior also affects the `agent.env` value, since it causes env var names to become lower-cased. Env vars are case-sensitive on non-Windows so modules might not expect it.)

### PR DESCRIPTION
**Pull request by arsing**
Fixes #1036

---

TODO:

- [x] ~~This probably breaks the use of env vars like `IOTEDGED_HOSTNAME` that could be used to override the config.yaml. Need to confirm.~~ It didn't break those which have lower-case keys already, like `hostname`, since the `Environment` source does still lower-case the keys it parses from the environment.

  It's probably acceptable since the ones that have mixed-case keys are deeply nested, and there were probably no users overriding them with env vars.

### CODE REVIEW

File: edgelet/edgelet-config/src/yaml_file_source.rs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+        let docs =
+            YamlLoader::load_from_str(&*s).map_err(|err| ConfigError::Foreign(Box::new(err)))?;
+
+        let mut docs = docs.into_iter();
+        let doc = match docs.next() {
+            Some(doc) => {
+                if docs.next().is_some() {
+                    return Err(ConfigError::Foreign(Box::new(std::io::Error::new(
+                        std::io::ErrorKind::Other,
```
Comment (kkdawkins) [inline edgelet/edgelet-config/src/yaml_file_source.rs]: Would InvalidInput be better if the user provided these yaml docs?

File: edgelet/edgelet-config/src/yaml_file_source.rs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        Box::new(self.clone())
+    }
+
+    fn collect(&self) -> Result<HashMap<String, Value>, ConfigError> {
+        let origin = match self {
+            YamlFileSource::File(path) => Some(path.to_string_lossy().into_owned()),
+            YamlFileSource::String(_) => None,
+        };
+
+        let s = match self {
```
Comment (kkdawkins) [inline edgelet/edgelet-config/src/yaml_file_source.rs]: nit: s is not a very readable variable name :)

File: edgelet/edgelet-config/src/yaml_file_source.rs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+                    result.insert(key, from_yaml_value(origin.as_ref(), value)?);
+                }
+            }
+        }
+
+        Ok(result)
+    }
+}
+
+/// Identical to https://github.com/mehcode/config-rs/blob/0.8.0/src/file/format/yaml.rs#L32-L68
```
Comment (kkdawkins) [inline edgelet/edgelet-config/src/yaml_file_source.rs]: Are there any license ramifications that arise from taking this code?

Comment (arsing) [reply edgelet/edgelet-config/src/yaml_file_source.rs]: We use it as MIT. It's okay as long as it's attributed, which is what that link does.

File: edgelet/edgelet-config/src/yaml_file_source.rs (hunk 1 above)
Comment (arsing) [reply edgelet/edgelet-config/src/yaml_file_source.rs]: I should probably just not be lazy and make an actual error type.

File: edgelet/edgelet-config/src/yaml_file_source.rs (hunk 2 above)
Comment (arsing) [reply edgelet/edgelet-config/src/yaml_file_source.rs]: Done.

**Pull request merged**